scrypto/src/math/decimal.rs

```diff
@@ -11,506 +10,591 @@ use sbor::rust::vec::Vec;
 
-    /// The fixed scale used by `Decimal`.
-    pub const SCALE: u32 = 18;
+                    pub const ZERO: Self = Self($zero);
 
-    pub const ZERO: Self = Self(0i128);
+                    pub const ONE: Self = Self($one);
 
-    pub const ONE: Self = Self(10i128.pow(Self::SCALE));
+                    /// Returns `$dec` of 0.
+                    pub fn zero() -> Self {
+                        Self::ZERO
+                    }
 
-    /// Returns `Decimal` of 0.
-    pub fn zero() -> Self {
-        Self::ZERO
-    }
+                    /// Returns `$dec` of 1.
+                    pub fn one() -> Self {
+                        Self::ONE
+                    }
 
-    /// Returns `Decimal` of 1.
-    pub fn one() -> Self {
-        Self::ONE
-    }
+                    /// Whether this decimal is zero.
+                    pub fn is_zero(&self) -> bool {
+                        self.0 == Zero::zero()
+                    }
 
-    /// Whether this decimal is zero.
-    pub fn is_zero(&self) -> bool {
-        self.0 == 0
-    }
+                    /// Whether this decimal is positive.
+                    pub fn is_positive(&self) -> bool {
+                        self.0 > Zero::zero()
+                    }
 
-    /// Whether this decimal is positive.
-    pub fn is_positive(&self) -> bool {
-        self.0 > 0
-    }
+                    /// Whether this decimal is negative.
+                    pub fn is_negative(&self) -> bool {
+                        self.0 < Zero::zero()
+                    }
 
-    /// Whether this decimal is negative.
-    pub fn is_negative(&self) -> bool {
-        self.0 < 0
-    }
+                    /// Returns the absolute value.
+                    pub fn abs(&self) -> $dec {
+                        $dec(self.0.abs())
+                    }
 
-    /// Returns the absolute value.
-    pub fn abs(&self) -> Decimal {
-        Decimal(self.0.abs())
-    }
+                    /// Returns the largest integer that is equal to or less than this number.
+                    pub fn floor(&self) -> Self {
+                        self.round(0, RoundingMode::TowardsNegativeInfinity)
+                    }
 
-    /// Returns the largest integer that is equal to or less than this number.
-    pub fn floor(&self) -> Self {
-        self.round(0, RoundingMode::TowardsNegativeInfinity)
-    }
+                    /// Returns the smallest integer that is equal to or greater than this number.
+                    pub fn ceiling(&self) -> Self {
+                        self.round(0, RoundingMode::TowardsPositiveInfinity)
+                    }
 
-    /// Returns the smallest integer that is equal to or greater than this number.
-    pub fn ceiling(&self) -> Self {
-        self.round(0, RoundingMode::TowardsPositiveInfinity)
-    }
+                    pub fn round(&self, decimal_places: u8, mode: RoundingMode) -> Self {
+                        assert!(decimal_places <= Self::SCALE.try_into().unwrap());
```

Comment:
Unless if we introduce 1024 bit integer types that have 308 zeros, that might overflow. And they will use them, because they can store 4 256 bits integers there for cheaper fees. So to be on the safe side I will do `u16`. (Actually we could name them `u64` too,  because they cost exactly the same amount of computations as rust does 64 bits calculations.)
I will resize them.

Reply:
Fixed in #8e279db5329317f6421a679bfa8074bf1524342f